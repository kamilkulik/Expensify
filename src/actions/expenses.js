const uuidv1 = require('uuid/v1');
import database from '../firebase/firebase';

//SYNCHRONOUS ACTIONS

/* 
component calls action generator
action generator returns an object
component dispatches object
redux store changes
*/

// ASYNCHRONOUS CODE

/*
component calls action generator
action generator returns function
component dispatches function (?) (requires redux middleware. Redux doesn't support dispatching functions out-of-the-box) : https://github.com/reduxjs/redux-thunk
function runs (has the ability to dispatch other actions and whatever it wants)
*/

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = {description, note, amount, createdAt};
        
        return database.ref(`users/${uid}/expenses`).push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    }
}

// REMOVE_EXPENSE
export const removeExpense = ( { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ( { id } = {}) => {
    return (dispatch, setState) => {
        const uid = setState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }))
            })
    }
};

// EDIT_EXPENSE

export const editExpense = ( id, updates ) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
    }
);

export const startEditExpense = (id, updates) => {
    return (dispatch, setState) => {
        const uid = setState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`)
        .update(updates)
        .then(() => {
            dispatch(editExpense(id, updates))
        })
    }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, setState) => {
        const uid = setState().auth.uid;
        const expenses = [];
        return database
            .ref(`users/${uid}/expenses`)
            .once('value')
            .then((snapshot) => {
                
                snapshot.forEach((childSnapshot) => {
                    const expense = {
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    }
                    expenses.push(expense);
                });

                dispatch(setExpenses(expenses));
            });
    }
}