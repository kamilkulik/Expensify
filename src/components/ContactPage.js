import React from 'react';

const ContactPage = () => (
    <div>
        This is my contact page
    </div>
);

export default ContactPage;