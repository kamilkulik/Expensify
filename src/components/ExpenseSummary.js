import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../utils/expensesTotal';
import numeral from 'numeral';

export class ExpenseSummary extends React.Component {
  render() {
    return (
      <div>
        <h1>
          {`Viewing ${this.props.expenseCount} 
          expense${this.props.expenseCount === 1 ? '' : 's'} 
          totalling ${numeral(this.props.expensesTotal / 100).format('$0,0.00')}`}
        </h1>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesTotal: expensesTotal(visibleExpenses),
    expenseCount: visibleExpenses.length
  }
}

export default connect(mapStateToProps)(ExpenseSummary);