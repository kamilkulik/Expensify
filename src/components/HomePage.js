import React from 'react'

const Home = () => (
    <div>
        <h1>Welcome</h1>
        <p>Thanks for stopping by. Take a look around!</p>
    </div>
);

export default Home;