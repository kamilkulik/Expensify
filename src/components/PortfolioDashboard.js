import React from 'react';
import { NavLink } from 'react-router-dom';

const portfolioPages = [
    {
        item: 'First portfolio item',
    },
    {
        item: 'Second portfolio item',
    },
    {
        item: 'Third portfolio item',
    },
]

const PortfolioDashboard = () => {
    return (
    <div>
        <h1>Checkout some of my work</h1>
        {portfolioPages.map((page, index) => {
            return (
                <NavLink key={index} to={`/work/${index+1}`} activeClassName="is-active">{page.item}</NavLink>
            )
        })}
    </div>
    )
};

export default PortfolioDashboard;