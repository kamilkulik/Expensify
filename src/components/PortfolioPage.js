import React from 'react';

const PortfolioPage = ({match, location}) => {
    return (
        <div>
        <div>
            <p>
                <strong>Match props:</strong>
                <code>{JSON.stringify(match, null, 2)}</code>
            </p>
            <p>
            <strong>Location props:</strong>
            <code>{JSON.stringify(location, null, 2)}</code>
            </p>
        </div>
        <div>

            <h3>This is my first portfolio page</h3>
        </div>
        </div>
    )};

export default PortfolioPage;