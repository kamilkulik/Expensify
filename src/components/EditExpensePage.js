import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (id, expense) => {
    this.props.editExpense(id, expense);
    this.props.history.push('/');
  }
  onClick = ({ id }) => {
    this.props.removeExpense( { id } );
    this.props.history.push('/');
  }
  render() {
    const id = this.props.match.params.id;
    return (
      <div>
        Editing the expense with id of {this.props.match.params.id}:
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button 
          onClick={this.onClick}
        >
        Remove
        </button>
      </div>
    );
  }
  
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense( data ))
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
