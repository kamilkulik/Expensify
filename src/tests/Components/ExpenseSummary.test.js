import expenses from '../fixtures/expenses';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import { shallow } from 'enzyme';
import React from 'react';
import expenseTotal from '../../utils/expensesTotal'

test('Should render ExpenseSummary', () => {
  const wrapper = shallow(<ExpenseSummary expensesTotal={expenseTotal(expenses)} expenseCount={expenses.length}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary expensesTotal={expenseTotal([expenses[1]])} expenseCount={[expenses[1]].length}/>);
  expect(wrapper).toMatchSnapshot();
});