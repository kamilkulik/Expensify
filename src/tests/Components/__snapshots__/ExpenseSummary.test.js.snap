// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseSummary 1`] = `
<div>
  <h1>
    Viewing 3 
              expenses 
              totalling $1,141.95
  </h1>
</div>
`;

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 1 
              expense 
              totalling $1,095.00
  </h1>
</div>
`;
